soc/nordic/sysbuild.cmake

```diff
@@ -0,0 +1,39 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+if(SB_CONFIG_VPR_LAUNCHER)
+  set(launcher_core "cpuapp")
+  string(REPLACE "/" ";" launcher_quals ${BOARD_QUALIFIERS})
+  list(LENGTH launcher_quals launcher_quals_len)
+  list(GET launcher_quals 1 launcher_soc)
+  list(GET launcher_quals 2 launcher_vpr)
+
+  string(REPLACE "cpu" "" launcher_vpr ${launcher_vpr})
+
+  if(launcher_quals_len EQUAL 4)
+    list(GET launcher_quals 3 launcher_variant)
+    if(launcher_variant STREQUAL "xip")
```

Comment:
The `xip` is a board variant, not a SoC defined CPU cluster, so this code does not belong inside a `soc/<vendor>/sysbuild.cmake` but inside board specific `sysbuild.cmake`.

Reply:
Those variants are common for all VPRs though: code can be executed from {M,R}RAM or SRAM, depending on the usecase. Only the memory layout changes in board specific files.
I think moving it to boards will introduce unnecessary duplication of code.

Reply:
From what I see, the only thing being decided here based on the board variant is which snippet to use.

Snippets are capable of selecting a different overlay for each board target, so I suggest combining `nordic-<vpr>` and `nordic-<vpr>-xip` into one. The sysbuild CMake code could then be simplified.

Reply:
Never mind, I forgot that both snippets target `cpuapp`. Sorry for the noise.

Reply:
> Those variants are common for all VPRs though: code can be executed from {M,R}RAM or SRAM, depending on the usecase. Only the memory layout changes in board specific files.
I think moving it to boards will introduce unnecessary duplication of code.

I understand, but with HWMv2 we need to keep the design a bit cleaner.
We're trying to instead extend the model as there is a need, instead of having pre-knowledge in SoC of what is defined for the board.

The `xip` variant is defined here: 
https://github.com/zephyrproject-rtos/zephyr/blob/215b4ddf989536c15d034cdf0941517b37927724/boards/nordic/nrf54l15pdk/board.yml#L7

This means another board could technically use another name, or not define it at all.
We have had a similar discussion regarding the `ns` variant which resulted in:
https://github.com/zephyrproject-rtos/zephyr/issues/70721

So if `xip` is a SoC variant, and the board should always define the variant, then we should have the option to define this at the SoC level.

But atm, `xip` in general is a config option, and not a board (or soc) variant, when looking at other boards.

Reply:
> Never mind, I forgot that both snippets target cpuapp. Sorry for the noise.

Not a bad idea.
Why strip the `xip` from the launcher board ?

That way only a single snippet is needed and depending on the board qualifier then the snippet decides which overlay and kconfig fragment to apply.
Thus the SoC code doesn't need to know anything about the qualifier.

Reply:
@masz-nordic @57300 @tejlmand ping

Reply:
To minimize the duplication involved in moving this to `boards`, I'd suggest selecting the snippet based on a promptless string config (defined in `soc`) whose value is populated in board-specific `Kconfig.sysbuild`:
```
config VPR_LAUNCHER_SNIPPET
        default "nordic-flpr" if BOARD_NRF54L15PDK_NRF54L15_CPUFLPR
        default "nordic-flpr-xip" if BOARD_NRF54L15PDK_NRF54L15_CPUFLPR_XIP
```

Reply:
to avoid detailed knowledge of board variants inside the SoC tree, could we make this code a bit more generic so that `xip` is not needed knowledge.

That will also allow for new drop-in snippets in future if new variants are appended.

See additional comments with code suggestions.